### DISCUSSION TITLE
Ошибка при рапсинге successful_payment

### DISCUSSION THREAD

**Discussion by YaDanilamaster** [reo7sp/tgbot-cpp#265] [Q&A] [answered]
Всем привет! Столкнулся с такой проблемой. 
После подтверждения платежа Телеграм отправляет подтверждение платежа, successful_payment. На update получаю от Телеграм вот такой HttpResponse:

{"ok":true,"result":[{"update_id":94549252,
"message":{"message_id":104,"from":{"id":2106570420,"is_bot":false,"first_name":"\u0454\u0438\u043d\u0458\u0448\u048b","username":"Krorodil_Gena","language_code":"ru"},"chat":{"id":2106570420,"first_name":"\u0454\u0438\u043d\u0458\u0448\u048b","username":"Krorodil_Gena","type":"private"},"date":1678978439,"successful_payment":{"currency":"RUB","total_amount":12000,"invoice_payload":"a","telegram_payment_charge_id":"6012453305_2106570420_664_7211157388846060692","provider_payment_charge_id":"ch_3MmIBlD2XCmrSOMK1h1aKdxa"}}}]}

На парсинге этого ответа программа падает вот тут:
https://github.com/reo7sp/tgbot-cpp/blob/master/src/TgTypeParser.cpp#L280
Это проблема только с тестовыми платежами?

**Comment by YaDanilamaster** [comment]
Внес изменения в TgTypeParser.cpp, указал значения по умолчанию для парсера Property Tree. Протестировал на платежах, ошибка не повторялась. Сделал Pull request с изменениями.

**Reply by Gorillza** [reply]
я так и не понял, у вас появилась возможность исполдьзования метода successfulpayment  после этих изменений или просто перестал падать Longpoll от такого json ?

**Comment by reo7sp** [comment, marked as answer]
В актуальной v1.12.3 эта проблема исправлена: старый Boost.PropertyTree parser удалён, `successful_payment` разбирается через nlohmann/json.

Для обработки платежа есть отдельный listener:

```cpp
bot.getEvents().onSuccessfulPayment(
    [](TgBot::Message::Ptr message, TgBot::SuccessfulPayment::Ptr payment) {
        std::cout << "chat_id=" << message->chat->id
                  << " amount=" << payment->totalAmount
                  << " currency=" << payment->currency << '\n';
    });
```

Событие приходит и для тестовых, и для реальных платежей. До него нужно своевременно отвечать на `pre_checkout_query`; `successful_payment` означает, что платёж уже завершён.
